lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -450,28 +489,121 @@ assignMemoryLayouts(llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
       // loadOp) has already be processed in a previous loop iteration. This
       // assumption is held by how loadOpsToIndirectionLevelAndUse recursively
       // collects loadOpToIndLevelAndUse using DFS.
-      if (loadToInfo.count(loadOp) == 0) {
+      if (loadsToPipeline.count(loadOp) == 0) {
         continue;
       }
     }
 
     // If we still don't have a shared encoding, try a "generic" shared
     // encoding.
-    if (!loadInfo.sharedEncoding && !isa<ttng::WarpGroupDotOp>(use)) {
-      loadInfo.sharedEncoding =
-          getSharedEncoding(op, /*isMMAV3=*/loadInfo.loadIsMMAV3)
-              .value_or(nullptr);
-      if (auto loadOp = dyn_cast<tt::LoadOp>(op)) {
-        loadInfo.blockedEncoding = getBlockedEncoding(loadOp, axisInfoAnalysis);
-      }
-    }
+    if (!hasSharedEncoding && !isa<ttng::WarpGroupDotOp>(use))
+      hasSharedEncoding = hasSharedEncodingHelper(op);
 
     // If that still didn't work, bail on pipelining this load.
-    if (!loadInfo.sharedEncoding) {
+    if (!hasSharedEncoding) {
+      continue;
+    }
+    loadsToPipeline.insert(op);
+  }
+  return loadsToPipeline;
+}
+
+static llvm::SmallVector<Operation *>
+getTransitiveUserInBlock(Operation *baseOp) {
+  llvm::SmallVector<Operation *> users;
+  DenseSet<Operation *> seen;
+  std::function<void(Operation *, Operation *)> dfs = [&](Operation *op,
+                                                          Operation *baseOp) {
+    if (!seen.insert(op).second)
+      return;
+    if (op != baseOp)
+      users.push_back(op);
+    if (isa<tt::LoadOp, tt::ExperimentalDescriptorLoadOp>(op)) {
```

Comment:
Don't we need to detect dot ops here too? If not, why?

Reply:
We are collecting all users including dot ops. Here is to stop the recursion once hitting the load ops. I think that is what we did in loadOpsToIndirectionLevelAndUse. In loadOpsToIndirectionLevelAndUse, once hitting a loadOp, the use will become the load op.

Reply:
Got it. But if the dot does not stop the recursion, does it mean that uses of the dot are also being considered transient uses of the load?

Reply:
> But if the dot does not stop the recursion, does it mean that uses of the dot are also being considered transient uses of the load?

Yes. For loadOpsToIndirectionLevelAndUse, if we have loadOp --> dot2 --> dot1, we will have [load, dot1] and [load dot2]. If the ForOp has num_stages attributes, the use can be a non-dot user.

Reply:
I don't think I follow the logic. If loadOp has immediate use in the same stage (let's say dot2 is in the same stage for whatever reason), and then dot1 is in later stage, loadOp is *not* pipelined, right? If I read this right, with this code we will still consider it pipelined?

Reply:
Oh I was talking about the current implementation of loadOpsToIndirectionLevelAndUse. And the logic here is trying to mimic what loadOpsToIndirectionLevelAndUse is doing.
If we want to generalize the logic to handle computation pipelining (i.e computation ops in different stages), we should also modify the logic here.

>  If loadOp has immediate use in the same stage (let's say dot2 is in the same stage for whatever reason), and then dot1 is in later stage, loadOp is not pipelined, right?
That is right.

If we want to generalize the logic in this patch, we should only check the immediate use of loadOp. In this specific case, only dot2 will be considered and since load and dot2 are in the same stage, load will not be pipelined. I can update the logic to generalize if that sounds good to you.

Reply:
Maybe I am misunderstanding, please tell me if what I say makes sense or not. I think this version introduces a problem that was not in there before the change: with this code, if we decide *not* to pipeline a load (so it is in the same stage as it's use, in my example above dot2), but in the kernel dot2 result is used in dot1, this code will still incorrectly determine that this load is pipelined.

Reply:
With the current implementation, I thought dot1 and dot2 will be in the same stage, all computation ops in scheduleLoads are put in numStages-1, loads and indirect loads are put in earlier stages.

> if we decide not to pipeline a load (so it is in the same stage as it's use, in my example above dot2), but in the kernel dot2 result is used in dot1, this code will still incorrectly determine that this load is pipelined.

If we assume dot2 is in the same stage as dot1, even if we consider dot1, this code will determine the load is not pipelined (because dot2 has the same stage as load, and dot1 has the same stage as dot2).

But I agree that we should not make the assumption. In that case, I think we just need to check the immediate uses (depth-1 uses)?

It is sad the other two landed patches are all reverted :[
I will follow up on the reverts. Thanks for reviewing! @pawelszczerbuk

Reply:
I guess whether or not dots are assigned to the same stage depends somewhat on the kernel. For example, if dot is followed by an arith op (let's say acc is scaled after the dot) that got assigned to different stage than the load and the dot, this code will produce incorrect results, I think.